Bluetooth: Controller: ISO Sync Receiver with ticks_slot_window

cvinayak opened the pull request:
Add implementation for ISO Sync Receiver to use ticks slot
window feature.

Added implementation to have ISO Sync Receiver resume its
subevents after being aborted by other overlapping events.

This will be a pat-on-the-back achievement, the BIG event can
start after an overlapping event, or yield to an overlapping event,
and resume back thereafter :)... for now a proof for sequential
packing implementations though.

cvinayak committed:
Bluetooth: Controller: ISO Sync Receiver with ticks_slot_window

Add implementation for ISO Sync Receiver to use ticks slot
window feature.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Introduce ticker reschedule with jitter

Introduce ticker reschedule with jitter so that role like
BIG can start after overlapping ACL and receive subsequent
subevents.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Resume feature for cancelled prepare

Add implementation to support resumption of cancelled
prepare.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: ISO Sync Receiver with resume feature

Added implementation to have ISO Sync Receiver resume its
subevents after being aborted by other overlapping events.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

Thalley commented on subsys/bluetooth/controller/ll_sw/lll_conn.h:
```diff
@@ -69,11 +69,20 @@ struct lll_conn {
 		struct {
 			uint8_t initiated:1;
 			uint8_t cancelled:1;
+			uint8_t forced:1;
+		};
+
+		struct {
+			uint8_t initiated:1;
+			uint8_t cancelled:1;
+			uint8_t forced:1;
 		} central;
```
What is the difference between these 2 structs now?

cvinayak replied:
unnamed struct is now used by a common code block and hence did not want to have a named reference as `central` or `peripheral`.

Thalley replied:
Suggest to put that in a comment, as that is non-obvious. Looks more like a merge conflict gone wrong :P

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_conn.c:
```diff
@@ -156,6 +170,17 @@ void lll_conn_abort_cb(struct lll_prepare_param *prepare_param, void *param)
+		 * back to central, otherwise let the supervision timeout
+		 * countdown be started.
+		 */
+		if (lll->role && (trx_cnt <= 1U)) {
```
If `role` is a specific value so that role == 0 == central and role == 1 == peripheral, then please make the role check more explicit. 

```suggestion
		if (lll->role == BT_HCI_ROLE_PERIPHERAL && (trx_cnt <= 1U)) {
```

The above assumes that the `role` field uses the HCI roles

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_sync_iso.c:
```diff
@@ -289,6 +292,35 @@ static int prepare_cb_common(struct lll_prepare_param *p)
+						&lll->data_chan_remap_idx);
+		}
+
+		drift_us %= lll->sub_interval;
+		drift_us = lll->sub_interval - drift_us;
```
I assume that `sub_interval` is also in `us`? 

But what is this doing? This looks like pure math magic to me - Can you add a comment what, why or how you are calculating a `drift_us` based on this?

cvinayak replied:
i am trying to calculate a remaining `sub_interval` of a remainder (modulus) of `sub_interval` of `drift_us`... now you see I am not good at explaining :-(...

Thalley replied:
I think my comment here is still valid :) It's not very intuitive code that should be commented

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_sync_iso.c:
```diff
@@ -397,6 +429,39 @@ static int prepare_cb_common(struct lll_prepare_param *p)
+}
+#endif /* CONFIG_BT_CTLR_SYNC_ISO_SLOT_WINDOW_JITTER */
+
+static int is_abort_cb(void *next, void *curr, lll_prepare_cb_t *resume_cb)
+{
+#if defined(CONFIG_BT_CTLR_SYNC_ISO_SLOT_WINDOW_JITTER)
+	if (!next) {
+		*resume_cb = resume_prepare_cb;
+
+		return -EAGAIN;
+	}
+#endif /* CONFIG_BT_CTLR_SYNC_ISO_SLOT_WINDOW_JITTER */
+
+	return -ECANCELED;
+}
```
Since you don't document internal functions like the `is_abort_cb` it's hard to follow and understand. 

Why are we here returning a pointer to a `resume_cb` while also returning an error, but not in most (all?) other cases?

cvinayak replied:
ok... it is not documented...i agree... i should...
0 - pre-emptee succeeds to continue radio use, and pre-emptor is informed
-EAGAIN - pre-emptee looses and requests a resumption later, and pre-emptor will be scheduled
-ECANCELED - pre-emptee looses and be aborted, and pre-emptor will be scheduled

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_sync_iso.c:
You are adding multiplying 2 uint8_ts and adding it to another uint8 and store the result in a uint8 without any checks for overflow. 

This looks dangerous to me, and suggest to either store the result in a uint16 and check the result, or add an assert for the other values to ensure that we don't get an overflow here.

cvinayak replied:
addressing here: https://github.com/zephyrproject-rtos/zephyr/pull/75638

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_sync_iso.c:
```diff
@@ -699,10 +769,15 @@ static void isr_rx(void *param)
 			lll->bn_curr = 1U;
 			lll->irc_curr++;
 
+			payload_offset = (lll->latency_event * lll->bn);
```
Ditto: Nothing to prevent an overflow here. 

This is, IMO, a general issue in the controller as values are rarely checked.

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_sync_iso.c:
```diff
@@ -1146,6 +1236,10 @@ static void isr_rx_done(void *param)
 	}
 	lll->payload_tail = payload_index;
 
+	if (latency_event--) {
+		goto isr_rx_done_latency_event;
+	}
```
What is the reasoning behind using `goto` here instead of a `while (latency_event--)`?

cvinayak replied:
it was a quick hack... will fix this

Thalley commented on subsys/bluetooth/controller/ticker/ticker.c:
```diff
@@ -1378,6 +1392,10 @@ void ticker_worker(void *param)
 #else  /* CONFIG_BT_TICKER_LOW_LAT ||
 	* CONFIG_BT_TICKER_SLOT_AGNOSTIC
 	*/
+		/* Pick the ticker expiry latency, to calculate ticks_at_expire
+		 */
```
```suggestion
		/* Pick the ticker expiry latency, to calculate ticks_at_expire */
```

Thalley replied:
Looks like you've still (mis?)configured your editor to use 80 instead of 100 line length with autoformatting.

Thalley commented on subsys/bluetooth/controller/Kconfig.ll_sw_split:
Additionally, it doesn't look like these new Kconfig options are ever tested

Thalley reviewed (changes requested):
This PR adds a lot of new code that is very hard to read due to the additional Kconfig options (`#if defined` generally makes things harder to read). 

Additionally the general lack of documentation for internal functions and their behavior and the long functions make it nearly impossible to actually understand what is going on.

cvinayak commented on subsys/bluetooth/controller/Kconfig.ll_sw_split:
```diff
@@ -742,8 +759,22 @@ config BT_CTLR_CENTRAL_RESERVE_MAX
 	  Note, currently this value is only used to space multiple central
 	  connections and not for actual ticker time reservations.
 
+config BT_CTLR_PERIPHERAL_RESERVE_MAX
```
No, that is peripheral ISO related. Where as this new one is related to normal peripheral ACL scheduling.

cvinayak reviewed (commented):
Thank you for reviewing, valuable inputs, and as always it takes me some time to refine the implementation, hope you retain your patience :-)

cvinayak commented:
rebased to ensure CI is passing, keeping this PR draft and will be sending smaller PRs with considerations to already provided review comments here.

Thalley commented:
@cvinayak what was the reason for closing this?

cvinayak commented:
This PR contained a mix of some fundamental changes to nRF specific Low Level scheduling (for which I will need new bsim tests, only tested using manual bsim samples); and, commits that fix/improve inter-operability with Sony Xperia and Samsung S24 phones. The fix/improvement commits are now in #77921 (which is also draft, as I need to complete handling of `is_abort_cb` mechanism for CIG events).

Also, I would like #78894 to be merged ahead.

To summarize the order:
1. #78894
2. #78096
3. #77921
4. #67231 or remaining commits of #75760 (both are enhancements)

Thalley commented on subsys/bluetooth/controller/Kconfig.ll_sw_split:
```diff
@@ -533,6 +540,16 @@ config BT_CTLR_SYNC_ISO_RESERVE_MAX
 	  disabled, then time reservation does not include the pre-transmissions
 	  and any Control subevents.
 
+config BT_CTLR_SYNC_ISO_SLOT_WINDOW_JITTER
+	bool "Wrap ISO Synchronized Receiver event around overlapping roles"
+	depends on BT_CTLR_SYNC_ISO && !BT_CTLR_SYNC_ISO_RESERVE_MAX
+	select BT_TICKER_EXT
+	help
+	  Wrap ISO Synchronized Receiver event around overlapping roles using
```
Might be useful to denote that it's the GAP roles here. There are many profiles with many roles, so it's non-obvious what roles are referred to here IMO

cvinayak committed:
tests: bsim: Bluetooth: Revert workaround for overlapping ACL

This reverts commit 490da9ec9613b5db547c1c87b66c5375a701e2b6.
This reverts commit 5f852998d046bf414f580cd1988cf5357a340c62.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_sync_iso.c:
```diff
@@ -192,9 +196,37 @@ static int prepare_cb(struct lll_prepare_param *p)
+
+	lll = p->param;
+
+	/* Deduce the latency */
+	lll->latency_event = lll->latency_prepare - 1U;
+
+	/* Update BIS packet counter to next value */
+	lll->payload_count += (lll->latency_prepare * lll->bn);
+
+	/* Reset accumulated latencies */
+	lll->latency_prepare = 0U;
+
+	/* Current window widening */
+	lll->window_widening_event_us += lll->window_widening_prepare_us;
```
Any chance of under or overflow in these operations?

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_sync_iso.c:
```diff
@@ -623,21 +627,36 @@ void ull_sync_iso_setup(struct ll_sync_iso_set *sync_iso,
 		 */
-		slot_us = lll->sub_interval * ((lll->nse * num_bis) - lll->ptc);
+		if (IS_ENABLED(CONFIG_BT_CTLR_SYNC_ISO_SLOT_WINDOW_JITTER)) {
+			slot_us = lll->sub_interval * lll->bn;
+			jitter_us = lll->sub_interval * lll->bn * (lll->irc - 1U);
+		} else {
+			slot_us = lll->sub_interval * ((lll->nse * num_bis) - lll->ptc);
+			jitter_us = 0U;
+		}
 
 	} else {
 		/* Time reservation omitting PTC subevents in interleaved
 		 * packing.
 		 */
-		slot_us = lll->bis_spacing * ((lll->nse - lll->ptc) * num_bis);
+		if (IS_ENABLED(CONFIG_BT_CTLR_SYNC_ISO_SLOT_WINDOW_JITTER)) {
+			jitter_us = lll->bis_spacing * num_bis * lll->bn * (lll->irc - 1U);
+			slot_us = lll->bis_spacing * num_bis * lll->bn;
+		} else {
+			slot_us = lll->bis_spacing * (lll->nse - lll->ptc) * num_bis;
+			jitter_us = 0U;
+		}
```
Similarly, with multiple multiplications: Are there any risk of overflow here?

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_scan_aux.c:
```diff
@@ -635,6 +635,11 @@ static int is_abort_cb(void *next, void *curr, lll_prepare_cb_t *resume_cb)
 	ARG_UNUSED(resume_cb);
 
+	/* Prepare being cancelled (no resume for scan aux) */
+	if (!next) {
```
```suggestion
	if (next == NULL) {
```

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_sync_iso.c:
```diff
@@ -198,9 +202,37 @@ static int prepare_cb(struct lll_prepare_param *p)
+	lll = p->param;
+
+	/* Deduce the latency */
+	lll->latency_event = lll->latency_prepare - 1U;
```
Can `latency_prepare` ever be `0`?

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_sync_iso.c:
```diff
@@ -259,10 +272,16 @@ static int prepare_cb_common(struct lll_prepare_param *p)
 	data_chan_id = lll_chan_id(access_addr);
 
+	/* Calculate the current event counter value */
+	event_counter = (lll->payload_count / lll->bn) - 1U;
```
Can `bn` ever be larger than `payload_count`?

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_sync_iso.c:
```diff
@@ -275,16 +294,56 @@ static int prepare_cb_common(struct lll_prepare_param *p)
+#if defined(CONFIG_BT_CTLR_SYNC_ISO_SLOT_WINDOW_JITTER)
+	uint8_t skipped = 0U;
+
+	if (p->ticks_drift) {
```
```suggestion
	if (p->ticks_drift != 0) {
```

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_sync_iso.c:
```diff
@@ -275,16 +294,56 @@ static int prepare_cb_common(struct lll_prepare_param *p)
 
+		drift_us = HAL_TICKER_TICKS_TO_US(p->ticks_drift);
+
+		/* FIXME: Add implementation to support interleaved packing */
+		LL_ASSERT(is_sequential_packing);
```
Is an assert the right option here? 

I'm wondering if we should just reject the sync request / terminate the BIG sync instead

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_sync_iso.c:
```diff
@@ -323,6 +382,26 @@ static int prepare_cb_common(struct lll_prepare_param *p)
+		/* Skipped subevents since last subevent reception */
+		irc_prev = lll->irc_curr;
+		lll->irc_curr = skipped + 1U;
+		skipped = lll->irc_curr - irc_prev;
```
I assume that `irc_curr` is always >= `irc_prev`?

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_sync_iso.c:
```diff
@@ -323,6 +382,26 @@ static int prepare_cb_common(struct lll_prepare_param *p)
 	}
 
+#if defined(CONFIG_BT_CTLR_SYNC_ISO_SLOT_WINDOW_JITTER)
+	if (skipped) {
```
```suggestion
	if (skipped != 0) {
```

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_sync_iso.c:
```diff
@@ -323,6 +382,26 @@ static int prepare_cb_common(struct lll_prepare_param *p)
+		skipped = lll->irc_curr - irc_prev;
+
+		/* Calculate the radio channel to use for subevent */
+		while (skipped--) {
```
```suggestion
		while (skipped-- != 0) {
```

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_sync_iso.c:
```diff
@@ -459,15 +538,74 @@ static int prepare_cb_common(struct lll_prepare_param *p)
 static int is_abort_cb(void *next, void *curr, lll_prepare_cb_t *resume_cb)
 {
+#if defined(CONFIG_BT_CTLR_SYNC_ISO_SLOT_WINDOW_JITTER)
+	if (!next) {
```
```suggestion
	if (next == NULL) {
```

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_sync_iso.c:
```diff
@@ -480,15 +618,26 @@ static void abort_cb(struct lll_prepare_param *prepare_param, void *param)
 
-		if (IS_ENABLED(CONFIG_BT_CTLR_BROADCAST_ISO_ENC)) {
-			const struct lll_sync_iso *lll = param;
+		if (false) {
```
The `if (false)` shouldn't be necessary here

Thalley reviewed (commented):
Some minor questions/comments

cvinayak committed:
tests: bsim: Bluetooth: Drift BAP Broadcast Assistant connection event

Update BAP Broadcast Assistant testing to have the ACL
connection drift over the BAP Broadcast Sink events.

Update tests to permit some amount of ISO Receive Errors.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak commented:
With the below changes:
```
diff --git a/subsys/bluetooth/controller/ll_sw/ull_adv_iso.c b/subsys/bluetooth/controller/ll_sw/ull_adv_iso.c
index f8f7cb35d64..67755d3ac9b 100644
--- a/subsys/bluetooth/controller/ll_sw/ull_adv_iso.c
+++ b/subsys/bluetooth/controller/ll_sw/ull_adv_iso.c
@@ -60,7 +60,7 @@
  * TODO: These could be a Kconfig option.
  */
 #define BT_CTLR_ADV_ISO_PTO_MIN         1U
-#define BT_CTLR_ADV_ISO_PTO_GROUP_COUNT 1U
+#define BT_CTLR_ADV_ISO_PTO_GROUP_COUNT 0U
 
 static int init_reset(void);
 static struct ll_adv_iso_set *adv_iso_get(uint8_t handle);
diff --git a/tests/bsim/bluetooth/audio/overlay-bt_ll_sw_split.conf b/tests/bsim/bluetooth/audio/overlay-bt_ll_sw_split.conf
index 5ccbca19665..d05a7c9280e 100644
--- a/tests/bsim/bluetooth/audio/overlay-bt_ll_sw_split.conf
+++ b/tests/bsim/bluetooth/audio/overlay-bt_ll_sw_split.conf
@@ -36,7 +36,7 @@ CONFIG_BT_CTLR_SYNC_ISO_SLOT_WINDOW_JITTER=y
 CONFIG_BT_CTLR_CENTRAL_RESERVE_MAX=n
 CONFIG_BT_CTLR_PERIPHERAL_RESERVE_MAX=n
 CONFIG_BT_CTLR_PERIPHERAL_ISO_RESERVE_MAX=n
-CONFIG_BT_CTLR_EVENT_OVERHEAD_RESERVE_MAX=y
+CONFIG_BT_CTLR_EVENT_OVERHEAD_RESERVE_MAX=n
 CONFIG_BT_CTLR_SLOT_RESERVATION_UPDATE=n
 CONFIG_BT_CTLR_SCAN_UNRESERVED=y
 CONFIG_BT_TICKER_NEXT_SLOT_GET_MATCH=y

```

Both BISes have a symmetric loss when a Peripheral ACL with 30 ppm drifts across the BIG event:

```
d_00: @00:03:27.118489  [00:03:27.118,469] <inf> bap_stream_rx: [19728|272|20000]: Incoming audio on stream 0x8511540 len 40, flags 0x09, seq_num 19999 and ts 207112241
d_00: @00:03:27.118489  [00:03:27.118,469] <inf> bap_stream_rx: [19728|272|20000]: Incoming audio on stream 0x851158c len 40, flags 0x09, seq_num 19999 and ts 207112241
```

cvinayak committed:
Bluetooth: Controller: Fix in-system ticker_start and ticker_stop tests

Fix in-system ticker_start and ticker_stop tests when
supporting BT_TICKER_EXT_EXPIRE_INFO.

Fixes commit d82c1dcaecae ("Bluetooth: Controller: In-system
ticker_start and ticker_stop tests").

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak commented on tests/bsim/bluetooth/audio/src/bap_stream_rx.c:
```diff
@@ -42,43 +45,54 @@ void bap_stream_rx_recv_cb(struct bt_bap_stream *stream, const struct bt_iso_rec
+	}
+
+	if (info->flags & BT_ISO_FLAGS_LOST && test_stream->valid_rx_cnt > 1U) {
+		LOG_ERR("ISO receive lost\n");
```
Done.

cvinayak committed:
Bluetooth: Controller: Fix Extended Scanning for slow CPUs

Add implementation to gracefully handle failure due to high
Radio ISR latencies due to higher CPU usage in LLL execution
context just before the handling of Radio IRQs. Higher CPU
time usage in LLL execution context can be due to slow CPUs,
example, nRF51x SoC Series with 16 MHz ARM Cortex-M0 CPU.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Fix ISO Receiver Sync Lost assertion

Fix ISO Receiver Sync Lost assertion due to successive
calls to stop ticker instance under conditions where there
are more than one ISO Sync LLL events enqueued in the LLL
prepare pipeline. This is visible when supporting LLL resume
for ISO Sync LLL events.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: ISO Sync Receiver with resume reservation

Added implementation to have ISO Sync Receiver resume ticker
to reserve time for next BIS when using sequential packing.
This is required so that any overlapping role that can
drift in its window be place after the next BIS subevent.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Fix window widening for PA Sync and ISO Sync

Fix window widening for Periodic Advertising Sync and ISO
Synchronized Receiver to correctly consider ULL skipped
and LLL prepare skipped events.

Relates to commit 2b4918545200 ("Bluetooth: Controller: Fix
window widening on connection update").

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: ISO Receiver aborts scan and scan aux overlap

Add implementation to cancel any overlapping scan and scan
aux instance with the ISO Receiver's event in progress.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Fix ISO Receiver subevent LLL scheduling

Fix ISO Receiver subevent LLL scheduling to consider radio
reception and jitter based on previous successful subevent
received in the event.

Without this fix later subevents that may have relative -2
us drift in the subevents with reference to the start of the
event do not get received.

Relates to commit e4b6751bbc92 ("Bluetooth: Controller:
Restrict ISO Sync Receiver subevent jitter").

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Fix out-of-bound memory access

Fix out-of-bound memory access calling
ull_sched_mfy_after_cen_offset_get() when being an
initiator.

Fix a redundant static scope declaration.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
tests: bsim: Bluetooth: audio: Enable tests that pass now

Enable tests that pass now and respective github issues can
be closed.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Fix missing window widening in time reservation

Fix missing window widening and permitted sleep clock jitter
in the ACL connection time reservations.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Fix ticker is_drift_in_window for lock step

Fix ticker is_drift_in_window to not take the interval of
the overlapping node, and hence prevent repeated drift at
every interval.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Improve Periodic Sync robustness

Improve Periodic Sync robustness.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Improve ISO Receiver robustness

Improve ISO Receiver robustness.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Improve ACL connection robustness

Improve ACL connection establishment by setting force flag
during the first 6 connection intervals when failing
reception or being preempted in the LLL prepare pipeline.

Relates to commit cadef5a64fa5 ("Bluetooth: Controller:
Introduce BT_CTLR_PERIPHERAL_RESERVE_MAX").

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: nRF5x: Tune counter compare minimum and latency

Tune counter compare minimum and latency.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Fix instant_latency for Central ISO setup

Fix missing implementation to handle instant_latency for
Central ISO setup.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Fix ticker start failure after immediate expiry

Fix ticker start from failing when bottom half does not run
for tickers that expire immediately, i.e. ticks_to_expire of
value 0, are requested to start again.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Fix mfy_aux_offset_get mayfly_enqueue assert

Fix mfy_aux_offset_get mayfly_enqueue assert when multiple
auxliary advertising sets enqueue while previous enqueue is
pending to be processed or in-progress in the ULL_LOW
context.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Use explicit margin for unreserved tickers

Update ticker implementation to use explicit ticks slot
margin for unreserved ticker collision resolution. This will
allow differentiating an unreserved ticker's needed free
space versus free space when the ticker is rescheduled.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Fix ticker ULL_HIGH and LLL operation counts

Fix ticker ULL_HIGH and LLL operation queue count for scan
auxiliary context scheduling and worst case preemption
requested LLL context enqueue elements.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Update BT_TICKER_EXT Kconfig default

Update BT_TICKER_EXT Kconfig default to y when there is
broadcaster, peripheral or extended advertising is enabled.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Add explicit BT_CTLR_SCAN_SLOT_WINDOW Kconfig

Add explicit BT_CTLR_SCAN_SLOT_WINDOW Kconfig to enable
placing primary scan window after overlapping states/
roles at end of its slot window so that it does not take
the interval of the overlapping states/roles.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Add explicit BT_CTLR_ADV_SLOT_WINDOW Kconfig

Add explicit BT_CTLR_ADV_SLOT_WINDOW Kconfig to enable
placing Advertising Primary PDUs after overlapping states/
roles at end of its slot window so that it does not take
the interval of the overlapping states/roles.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Fix missing BT_CTLR_PERIPHERAL_RESERVE_MAX

Fix missing conditional compilation under connection update.

Relates to commit cadef5a64fa5 ("Bluetooth: Controller:
Introduce BT_CTLR_PERIPHERAL_RESERVE_MAX").

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Fix ticker for drift in window to yield

Fix ticker implementation for tickers with drift in window
to yield when overlapping with other ticker nodes.

Relates to commit 9df4b548ba32
Fixes commit b838b17755b2

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: Controller: Fix missing ticks_drift reset

Fix missing ticks_drift value reset when ticker is inserted
back at its next periodic interval. Without this rescheduled
expiry had the value of ticks_slot_window as the ticks_drift
value.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>
